streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java

```diff
@@ -1103,32 +1156,67 @@ public void process(final ProcessorSupplier<? super K, ? super V> processorSuppl
                                                    final KStream<K1, V2> other,
                                                    final ValueJoiner<? super V1, ? super V2, ? extends R> joiner,
                                                    final JoinWindows windows,
-                                                   final Joined<K1, V1, V2> joined) {
+                                                   final Joined<K1, V1, V2> joined,
+                                                   final Materialized<K1, ?, WindowStore<Bytes, byte[]>> materialized) {
```

Comment:
The critical point for topology compatibility in `KStreamImplJoin#join` is that we continue to increment the counter five times in this section of code, even when users provide the name for the join processor, repartition topic (via `Joined`) and the state stores (via `Materialized`).